File: generate-parentheses/generate-parentheses.cpp
class Solution {
    vector<string>ans;
public:
    vector<string> generateParenthesis(int n) {
        generate(0, 0, n, "");
        return ans;
    }

    void generate(int open, int close, int n, string s)
    {
    if (s.size() == 2 * n)
        ans.push_back(s);
    if (open < n)
        generate(open + 1, close, n, s + "(");
    if (close < open)
        generate(open, close + 1, n, s + ")");
    }
};

File: generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n):
        self.ans = []
        s = ""
        self.generate(s, n, n)
        return self.ans

    def generate(self, s, o, c):
        if o == 0 and c == 0:
            self.ans.append(s)
            return
        if o > 0:
            s += '('
            self.generate(s, o - 1, c)
            s = s[:-1]
        if c > 0:
            if o < c:
                s += ')'
                self.generate(s, o, c - 1)
                s = s[:-1]
